New schema

aviralv opened the pull request.

aviralv committed:
Setting the new metainfo structure

aviralv committed:
Added more fields from the JSON schema

aviralv committed:
Lists work in schematic

aviralv committed:
Support for pandas, adding data to nomad archive

aviralv committed:
Added Data headers

aviralv committed:
Added data files for testing

aviralv committed:
Removed printing header of dataframe

aviralv committed:
Fixed multiple elements in a list.

aviralv committed:
Checking if the dataset exists

aviralv committed:
Using os.path, fixed string manipulation

aviralv committed:
Fixed magic number 15, reading dataset file header

aviralv committed:
Cleaned up author generated section

aviralv committed:
Added spectrum instead of numerical values array

aviralv committed:
Added pint library

aviralv committed:
Moving 'edges' to experiment

aviralv committed:
Added units to quantities

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
+        #Check that the msa file exists
+        
+        dirpath = filepath[:filepath.rindex('/')]
+        if os.path.exists(dirpath):
```
Using glob was fine, right, and shorter. The only problem was that you blindly expected glob to return a non empty list.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
+
+        #Check that the msa file exists
+        
+        dirpath = filepath[:filepath.rindex('/')]
```
You have to use `os.path` this. Manually manipulating paths is less expressive and not platform independent (think windows paths for example).

aviralv replied:
Resolved this. I'll be more mindful of platform independence.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
+        if os.path.exists(dirpath):
+            for i in os.listdir(dirpath):
+                if i.endswith('.msa'):
+                    data_file_path = os.path.join(dirpath, i)
```
The variable naming is not consistent: `...file_path` vs `filepath`

aviralv replied:
filepath is the variable that is inherited by the function. I have now renamed the dataset filepath to be dataset_filepath. That should be consistent with our naming.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
+                    data_file_path = os.path.join(dirpath, i)
+        
+        #Read header from the msa file
+        with open(data_file_path, 'rt') as f:
```
`data_file_path` is not set, if there is no '.msa' file.

aviralv replied:
Fixed by checking if the variable exists, then the script creates the dataset under the sebsection data

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
+                skip_row_counter += 1
+                header_data_file.append(previous_line.strip().lstrip('#').split())
+        
+        #Extract units from the header file
```
The comment is misleading, there is no "header file"

aviralv replied:
Header of the dataset file.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
+            skip_row_counter = 0
+            header_data_file = []
+            while line.startswith('#'):
+                previous_line = line
```
Why the previous_line? You can reply read the new line at the end of the loop.

aviralv replied:
Yeah. That was unnecessary. I've made the required changes.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
+        with open(data_file_path, 'rt') as f:
+            line = f.readline()
+            skip_row_counter = 0
+            header_data_file = []
```
That is a bad name, what is a "header data file"?

aviralv replied:
header_dataset

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
+        #Export the dataset to the archive
+        # numerical_value = data.m_create(NumericalValues)
+        # numerical_value.data_values = df.to_numpy()
+        ureg = pint.UnitRegistry()
```
You have to use nomads registry `nomad.units.ureg`. This is necessary to have consistent natural constants and unit conversions throughout nomad.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
+
+        spectrum = data.m_create(Spectrum)
+        spectrum.n_values = len(df)
+        spectrum.energy = Q_(df[0].to_numpy(), x_units)
```
That is the right idea. But I think you could also write something like `df[0].to_numpy() * ureg(x_units)`; `<value> * <unit>` seems to be more "physical" than `Q_(<value>, <unit>). It is nitpicking, I know.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
-                    archive.section_metadata.references.append('; '.join([
-                        str(value).strip() for value in reference.values()
-                    ]))
+            sample.elements = elements
```
Not you fault, but thats ugly. Is there really "json in json" in the metadata files? Maybe we should improve the crawl script.

aviralv replied:
There isn't json in json. I copied this part from the existing code.

markus1978 replied:
We can try to remove this, when we run parsing on everything. This `if: json.loads` was probably put in for a reason. You could download our eelas data from NOMAD (or better retrieve an updated set from eels with the crawl script) and test parse everything. Eeals isn't too big.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
+        instrument = metadata.m_create(Instrument)
+        instrument.source_label = file_data['microscope']
+        device_settings = instrument.m_create(DeviceSettings)
+        device_settings.device_name = file_data['microscope']
```
I am not sure how consistent all these keys are actually available. Running it over all of EELS will show.

aviralv replied:
Yeah. These things were available in the two test cases that I was using. Is there an actionable step here for now?

markus1978 replied:
Yes, test the parser on the whole eelsdb. If it turns out that the data is not consistent, we have to make the array access safer.

aviralv replied:
I pulled the entire eelsdb and resolved all the key bindings for the metadata files.
I, subsequently, found a few cases where the dataset is not consistent. I resolved the minor error with different separators in the files. But there are still a few problem cases.
These either don't have any header in the dataset file, or the units are undefined in the header of the dataset.

Should we do something about that?

markus1978 replied:
This is very the dirty business of data integration starts. Not having the data or having possibly flawed data. NOMAD usually does the second by carefully applying sensible defaults. E.g., you simply assume that it is eV.

aviralv replied:
I can put that to be 'eV' in case of EELS Data, but there might be situations where it could be 'keV' or 'meV' and then we're off.

markus1978 replied:
I would still take eV as default and produce a warning via our logger.

aviralv replied:
Alright, I've set it to eV by default. I'll add a logger warning.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
+        #Populating the data header; the information isn't in the metainfo file
+        #Data Header
+        
+        for i in range(len(df.columns)):
```
This is unnecessary. We put the spectra definitions in the metainfo to avoid this. Not need for DataHeaders.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +48,186 @@ def __init__(self):
+                data_header.unit = y_units
+        
+
+        # experiment = archive.m_create(Experiment)
```
Don't commit commented code. We use git to remember old versions.

markus1978 commented on eelsdbconverter/metainfo.py:
```diff
@@ -0,0 +1,113 @@
+class Spectrum(MSection):
+  n_values = Quantity(type=int)
+  energy = Quantity(type=np.dtype(np.float64), shape=['n_values'], unit='eV', description='This column contains the range of energies of the spectrum')
+  count = Quantity(type=np.dtype(np.float64), shape=['n_values'], unit=' ', description='The count at each energy value, dimensionless')
```
Just provide no unit instead of ' '

markus1978 commented on eelsdbconverter/metainfo.py:
```diff
@@ -0,0 +1,113 @@
+
+class Spectrum(MSection):
+  n_values = Quantity(type=int)
+  energy = Quantity(type=np.dtype(np.float64), shape=['n_values'], unit='eV', description='This column contains the range of energies of the spectrum')
```
The metainfo uses SI-units, e.g. J instead of eV. Using the pint quantity in the parser will automatically convert.

markus1978 replied:
The descriptions should be sentences including a ".". These are just energy values, and they should not be qualified as columns, rows, arrays, or whatever. E.g., "The energy values of the spectrum."

markus1978 commented on eelsdbconverter/metainfo.py:
```diff
@@ -0,0 +1,113 @@
+    section_data_header = SubSection(sub_section=DataHeader, repeats=True)
+
+class Data(MSection):
+    repository_name = Quantity(
```
Maybe we need a metadata sections for data? Everything here except section_spectrum is not really "the data", but metadata about how to find it, preview it, etc.

aviralv replied:
That was the whole point to split the whole measurement into data and metadata. section_spectrum is the numerical data, but everything else is the metadata to this data.

I'm not sure if I follow your point here.

markus1978 replied:
The quantities `repository_url`, `preview_url`, `entry_repository_url`, `repository_name` are not really data and should not be in the section data. You define section data to hold the actual data. The original intend behind this section was different and more like "metadata that is data related". Therefore these could/should move somewhere else. Maybe `measurement.metadata.origin` or something. `preview_url` might be arguable if this actually belongs in data?

aviralv replied:
Yeah, I agree. I missed the point earlier. I could move these currently to `AuthorGenerated` because very loosely the database in this case is the source of the data. And these are miscellaneous field? I'm a bit hesitant to add more new subsections. 
Or would you still recommend a different section header? Like `measurement.metadata.origin`

markus1978 replied:
Again, i don't like `AuthorGenerated` as a section. It is fine to have `Author` (when the data describes the Authors, name, affiliation, etc.). Otherwise, stuff in AuthorGenerated should move to where it thematically belongs. That a quantity is generated by a human, can be marked otherwise (e.g. via Metainfo Category). 

Following this, I would suggest a different section, like `origin`. The only questionable quantity is `preview_url`. I guess you find arguments for having this in `measurement.data`, or `measurment.metadata`, or `metasurement.metadata.origin`?

aviralv replied:
The reason to have `AuthorGenerated` as a prominent section is because this is the section where quantities such as `internal_project_id`, `internal_sample_id `would be expected to go. The other items in the list are expected to be human generated, such as `author_notes`. 
Currently, we need a section like `origin` as you have pointed out because we're reading off data from a database, but is that expected to be a usual case? If that is expected to be the case, then I can create a section like `origin`.

markus1978 replied:
We are discussing now two independent topics. 

There will be different types of origins: external databases (e.g. EELS), an oasis (e.g. XPSHub), an 3rd party repository (e.g. Xenodo). There are probably more. Consider that you might only have the metadata at hand. Maybe we even need subsections to `measurment.metadata.origin`. 

I dislike the idea of having a plain bucket for everything that an author generates. The sections should describe a common theme for what their quantities are about and not how they were obtained. If it is a sample id it belongs in sample, if it is an author note, its either a note about the sample, instrument, or the measurement in general. What if the sample id is not "author generated"? Building the containment hierarchy on different criteria (topic and method of acquisition) raises questions about where to put quantities. E.g, does sample id belong in sample or author generated?

We could generalize and add certain common quantities to all metadata sections (internal_id, comment, references)?

aviralv replied:
I agree to the points you made. I'll create a new section `origin` and adding certain quantities to all Metadata sounds like the best solution. I'll change the metainfo.py to reflect these.

aviralv committed:
Incorporated suggestions from Markus

aviralv committed:
Copying code styles to eels parser

aviralv committed:
Formatting changes

aviralv committed:
Fixed metainfo.py for stylechecks

aviralv committed:
Style changes to __init__.py

aviralv committed:
Fixed if dataset (msa file) doesn't exist

aviralv committed:
Style changes

markus1978 reviewed (commented):
See my answers to your comments.

aviralv committed:
Updated .gitignore to ignore all spectral data files

aviralv committed:
Moved data origin to AuthorGenerated

aviralv committed:
Ensured that the keys exist; problem cases are in the tests directory

The problem cases either don't have any header in the dataset file, or the units are undefined in the header of the dataset.

aviralv committed:
Fixed issue with phantom columns on changing to multiple separator types; also ensured that the default unit for energy is 'eV'

aviralv committed:
Created section Origin

aviralv committed:
Created corresponding entries to Origin in __init__.py

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+        # Create a measurement in the archive
+        measurement = archive.m_create(Measurement)
+
+        """
```
That is super verbose. Also  it is not treated as a comment, it is an actual statement that python evaluates every time the function runs. Just do:
```
# This is a comment, and even if it is multi
# line, its fine.
```

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+        # Check that the msa file exists
+
+        dirpath = os.path.dirname(filepath)
+        for i in os.listdir(dirpath):
```
You should to the glob, just check for empty result, e.g.:
```python
dataset_filepath = next(iter(glob(os.path.join(dirpath, '*.msa'))), None)
```

Maybe even check if there is more than one. Just choose one and produce a warning? Maybe also a warning on no .msa file?

aviralv replied:
The last update should produce a warning for no *.msa file, but it doesn't check for a possibility of multiple *.msa files

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+            # Read header of the dataset(msa file)
+            with open(dataset_filepath, 'rt') as f:
+                line = f.readline()
+                skip_row_counter = 0
```
You don't need skip_row_counter, pandas read_csv has a comment parameter to enable ignoring of lines starting with #.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+            with open(dataset_filepath, 'rt') as f:
+                line = f.readline()
+                skip_row_counter = 0
+                header_dataset = []
```
You don't need to construct header_dataset, if you do the unit reading from next loop right here.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+                skip_row_counter = 0
+                header_dataset = []
+                while line.startswith('#'):
+                    header_dataset.append(line.strip().lstrip('#').split())
```
use a regexp (can be used to extract units right away)

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+
+            # Extract units from the header of the dataset file
+            for item in header_dataset:
+                if item[0].lower() == 'xunits':
```
use a regexp to extract this.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+                        x_units = 'eV'
+
+            # Read the dataset from the msa file
+            df = pd.read_csv(dataset_filepath,
```
don't need skiprows, use `comment='#'`

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+            # Read the dataset from the msa file
+            df = pd.read_csv(dataset_filepath,
+                             header=None, skiprows=skip_row_counter,
+                             skipfooter=1, engine='python', sep=", | |,|\t")
```
Using `'` over `"` is the right choice, but be consistent. (Also for doc strings)

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+                             skipfooter=1, engine='python', sep=", | |,|\t")
+
+            # Export the dataset to the archive
+            ureg = nomad.units.ureg
```
Do this with an import `from nomad.units import ureg` at the top.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+            spectrum.n_values = len(df)
+            try:
+                x_units
+            except NameError:
```
Wow, no. Initialize `x_units = None` before and do this with `if x_units is not None`.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+                x_units
+            except NameError:
+                x_units = 'eV'
+                logger.info('WARNING!!! Manually set energy units to eV!')
```
User `logger.warn()`; no `WARNING!!!` necessary. Also drop the '!'s. *Manually* and *set* is poor word choice: *energy unit unknown, assuming eV by default*

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+                logger.info('WARNING!!! Manually set energy units to eV!')
+            if 'undefined' in x_units:
+                x_units = 'eV'
+                logger.info('WARNING!!! Manually set energy units to eV!')
```
Should go in one if: `if x_units not None or 'unknown' in x_units:`

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+        device_settings = instrument.m_create(DeviceSettings)
+        device_settings.device_name = file_data['microscope']
+        device_settings.max_energy = file_data['max_energy']
+        if file_data.get('min_energy') is not None:
```
I think `if 'min_energy' in file_data:` is easier to read (and might also be cheaper).

aviralv replied:
I followed the coding style used in the previously written parser here. Do you want me to update it nevertheless? I can do it for all the definitions that I've done this way.

markus1978 commented on eelsdbconverter/__init__.py:
```diff
@@ -49,60 +44,129 @@ def __init__(self):
+        device_settings.dark_current = file_data['darkcurrent']
+
+        # Author Generated
+        author_generated = metadata.m_create(AuthorGenerated)
```
Again, the author info should be in a section Author. (For external dbs) This should be a subsection of origin. Other user information (uploader, coauthors are treated by nomad and stored somehwere else). 

The description belongs to section measurement. No AuthorGenerated pls. We can have a category to denote "author generated": 

```python
class UserProvided(metainfo.Categroy):
    pass

class Measurement(metainfo.Section):
    description = Quantity(type=str, categories=[UserProvided]))
```

markus1978 replied:
Most sections should get their own UserProvided quantities. Measurement, Sample, Instrument, etc. all should have something like notes, comment?

aviralv replied:
I created a category like you suggested. I agree that it makes more sense to do it this way. This was the desired result of having a section called `AuthorGenerated`. Just check that I've done this correctly.

aviralv committed:
Updates to feedback from Markus; regex to read x_units pending

aviralv committed:
Used Regex to exxtract x_units

markus1978 commented:
Could you also add the requirements.txt from the example parser. Otherwise, the actions won't run because the pycodestyle, plint, etc. dependencies are not installed.

aviralv committed:
Adding requirements.txt

aviralv commented:
Done

aviralv committed:
Adding python actions

aviralv committed:
Testing update for EELS

aviralv committed:
Update python-actions.yaml

aviralv committed:
Update python-actions.yaml

aviralv committed:
Update python-actions.yaml

aviralv committed:
Changed import order of python libraries

aviralv committed:
Added .pylintrc

aviralv committed:
Updated logger.warn to logger.warning;

aviralv committed:
Update requirements.txt

aviralv committed:
Created tests for the parser

aviralv committed:
Update test_parser.py

aviralv committed:
Update test_parser.py

aviralv committed:
Update test_parser.py

aviralv committed:
Update test_parser.py
